drivers/i3c/i3c_renesas_imx3112.c

```diff
@@ -0,0 +1,619 @@
+	if (err) {
+		return err;
+	}
+
+	err = i3c_renesas_imx3112_set_channel(channel_config->mux, channel_config->channel_mask);
+	if (err) {
+		goto end_trans;
+	}
+	/* Change the device bus from the mux to the controller to redirect the transaction */
+	target = i3c_device_find(bus_dev, &pid);
+	if (target == NULL) {
+		LOG_ERR("Mux dev %s failed to find target: 0x%llX", channel_dev->name,
+			(uint64_t)pid.pid);
+		return -ENXIO;
+	}
+	err = i3c_transfer(target, msgs, num_msgs);
+
+end_trans:
+	k_mutex_unlock(&data->lock);
+	if (err) {
+		LOG_ERR("Mux dev %s transfer failed with error %d", channel_dev->name, err);
+	}
+	return err;
+}
+
+static int i3c_renesas_imx3112_configure(const struct device *dev, enum i3c_config_type type,
+					 void *config)
+{
+	return -ENOSYS;
+}
+
+static int i3c_renesas_imx3112_i2c_configure(const struct device *dev, uint32_t dev_config)
+{
+	return -ENOSYS;
+}
+
+static struct i3c_device_desc *i3c_renesas_imx3112_device_find(const struct device *channel_dev,
+							       const struct i3c_device_id *id)
+{
+	const struct i3c_renesas_imx3112_channel_config *channel_config = channel_dev->config;
+
+	return i3c_dev_list_find(&channel_config->device_list, id);
+}
+
+static int i3c_renesas_imx3112_channel_initialize(const struct device *channel_dev)
+{
+	LOG_DBG("Initalizing mux channel %s", channel_dev->name);
+	const struct i3c_renesas_imx3112_channel_config *channel_cfg = channel_dev->config;
+	const struct i3c_renesas_imx3112_mux_config *mux_cfg =
+		get_mux_config_from_channel(channel_dev);
+	int err;
+
+	if (!device_is_ready(channel_cfg->mux)) {
+		LOG_ERR("I3C mux root %s not ready", channel_cfg->mux->name);
+		return -ENODEV;
+	}
+
+	/* Test channel selection */
+	err = i3c_renesas_imx3112_set_channel(channel_cfg->mux, channel_cfg->channel_mask);
+	if (err) {
+		return err;
+	}
+
+	if (mux_cfg->transmission_mode == MODE_I3C) {
+		/* Run a SETAASA on the channel to switch all children to I3C mode */
+		err = i3c_ccc_do_setaasa(mux_cfg->bus);
+		if (err) {
+			return err;
+		}
+	}
+	LOG_DBG("Mux channel %s initalization complete", channel_dev->name);
+	return 0;
+}
+
+/* Additional initialization steps needed to set the mux to I3C mode */
+static int i3c_renesas_imx3112_mux_configure_i3c_mode(const struct device *dev)
+{
+	LOG_DBG("Setting mux %s to I3C mode", dev->name);
+
+	const struct i3c_renesas_imx3112_mux_config *cfg = dev->config;
+	struct i3c_renesas_imx3112_mux_data *data = dev->data;
+	int err;
+
+	/*
+	 * Need to grab the pointer to the I3C device descriptor or I2C device descriptor
+	 * before we can configure the mux
+	 */
+	data->mux_desc_i3c = i3c_device_find(cfg->bus, &cfg->dev_id_i3c);
+	if (data->mux_desc_i3c == NULL) {
+		LOG_ERR("Cannot find I3C device descriptor");
+		return -ENODEV;
+	}
+
+	/*
+	 * Run a SETAASA on bus to switch to I3C mode. If the mux is already in I3C mode
+	 * this CCC will be ignored
+	 */
+	err = i3c_ccc_do_setaasa(cfg->bus);
```

Comment:
I'm afraid this could have consequences with initialization of other devices on the bus as this is a broadcast message where it cause ALL devices on the bus (if SETAASA supported) to load their static address as their dynamic address regardless of what is intended from the device tree. I don't have the IMX3112 data sheet in front of me... but are you sure this is the only way for this to "switch" to I3C mode. That CCC is just one of the ways of assigning a DA. The controller of the bus this is attached to should be the one already assigning a DA to this mux which should initialized before this which should already be doing as apart of it's init func.
This really does not seem right.

Reply:
Thanks Ryan for taking a look :)

The reason why I don't set dynamic addresses for each of the child devices in the driver code is that all the child devices we use in the mux just use their SA's as DA's, so we never had to set DA's individually (i.e use ENTDAA). I'll investigate how to do other methods of address assignment and see if i can adjust the driver code. 

 I agree that setting the device to I3C should be left to the I3C controller, so we can get rid of this of this SETAASA.

Reply:
Yeah... it should, but it appears this may run into a limitation of the MACRO that gathers up all the i3c devices on the bus. I ran in to this issue with the tca9546. Which I wrote about there... https://github.com/zephyrproject-rtos/zephyr/issues/51828

I still haven't figured out the macrobatics yet to solve it :(

Reply:
Ah yeah, thank you for adding the attach/detach api https://github.com/zephyrproject-rtos/zephyr/pull/52035. Its unfortunate we haven't found a macrobatic expert to get us a macrobatic i2c/i3c tree yet :/.

As a workaround for now, we have been using the following design pattern:

```
&i3c_controller {
      lps22hh0_mock: lps22hh_mock@10 {
              compatible = "vnd,i3c-device";
              reg = <0x10 0xabcd 0x12345010>;
      };

      /* The imx3112 node must be a child of an i3c controller */
      mux: imx3112@77 {
          compatible = "renesas,imx3112-i3c";
          reg = <0x77 0xabcd 0x02345678>;
          status = "okay";
          #address-cells = <1>;
          #size-cells = <0>;
          mux_i3c@0 {
              compatible: "renesas,imx3112-channel"
              reg = <0>;
              #address-cells = <3>;
              #size-cells = <0>;
              lps22hh0: lps22hh@10 {
                      compatible = "st,lps22hh";
                      reg = <0x10 0xabcd 0x12345010>;
              };
          };
          mux_i3c@1 {
            compatible: "renesas,imx3112-channel"
              reg = <1>;
              #address-cells = <3>;
              #size-cells = <0>;
              lps22hh1: lps22hh@10 {
                      compatible = "st,lps22hh";
                      reg = <0x10 0xabcd 0x12345010>;
              };
          };
      };
};

```
Where the vnd,i3c-device can just be a stub driver.

Obviously this works best if we have the same addresses in each channel, however.

Reply:
Do you think we can resolve this? I am unsure if there's a better way to switch to I3C for this mux's children, since the mux does not support ENTDAA.

Reply:
@XenuIsWatching I was hoping you could take a final look, thanks :)

Reply:
sorry, I don't have the data sheet for this part so I can't really dig into what it supports... (it sounds odd that something can not support ENTDAA it is required for all i3c devices)
I really don't know what mean by "switch to I3C".... do you just mean assign it's DA?

Ideally, how I'd like to see this resolve is for additional Python code to be created to allow for additional macros that would allow the I3C macros to gather the devices in the mux... but then there is the issue of how the address table would handle duplicate SA

I don't believe sending SETAASA is right all here... That is an optional CCC command that devices can support, so it working with all devices can't be guaranteed here.

Reply:
> Maybe [...] @mbolivar-ampere (so many handles here...) has an idea about the python code https://github.com/zephyrproject-rtos/zephyr/issues/61928

This is the correct current account. Sorry for the churn.

I think we talked about this in person. There's no issue conceptually with adding a `DT_FOREACH_DESCENDANT(node_id, fn)` macro that expands to calling `fn` on each descendant of `node_id` (although you might want to have some extra macros to control visitation order -- depth first, breadth first, etc. -- depending on the use case).

You would do this by adding to `write_children(node)` in gen_defines.py using the edtlib.py API, adding the macro itself and doxygen comments to zephyr/devicetree.h, and adding test cases in tests/lib/devicetree/api.

I don't have any plans to implement this but of course I would review patches adding it.

Reply:
Really sorry for the delay, I was busy with non-zephyr-related matters, and I forgot about the PR.

I removed the SETAASA for now. I think each child driver in the mux can do its own switch to I3C if necessary.
As to the IMX3112 spec sheet/programming guide, I don't think its public yet, although I could ask the support team about that.

I think that being able to expand DT_FOREACH_DESCENDANT(node_id, fn) to do N levels down the tree would be cool, I just don't think I have the Zephyr knowhow to do this :)